File: previous-gen-apps/voice-kraliki/frontend/src/lib/api/aiServices.ts
/**
 * AI Services API Client
 *
 * Provides typed API calls to backend AI services:
 * - Transcription management
 * - Call summarization
 * - Agent assistance
 * - Sentiment analysis
 */

const API_BASE = import.meta.env.VITE_BACKEND_URL || 'http://localhost:8000';

// ============================================================================
// Types
// ============================================================================

export interface TranscriptionConfig {
	language?: 'en' | 'es' | 'cs' | 'de' | 'fr';
	enable_interim_results?: boolean;
	enable_speaker_detection?: boolean;
	enable_punctuation?: boolean;
}

export interface TranscriptionSegment {
	id: string;
	session_id: string;
	text: string;
	speaker: 'agent' | 'customer' | 'system';
	confidence: number;
	is_final: boolean;
	timestamp: string;
}

export interface SummarizationConfig {
	language?: string;
	include_action_items?: boolean;
	include_sentiment?: boolean;
	include_topics?: boolean;
	max_summary_length?: number;
}

export interface CallSummary {
	session_id: string;
	summary: string;
	key_points: string[];
	action_items: Array<{
		description: string;
		priority: 'low' | 'medium' | 'high';
		due_date?: string;
		assigned_to?: string;
	}>;
	outcome: 'resolved' | 'escalated' | 'pending' | 'callback_required';
	customer_sentiment: string;
	agent_performance_score: number;
	topics: string[];
	duration_seconds: number;
	timestamp: string;
}

export interface AssistanceConfig {
	enable_suggestions?: boolean;
	enable_knowledge_base?: boolean;
	enable_compliance?: boolean;
	enable_coaching?: boolean;
	suggestion_threshold?: number;
	max_suggestions?: number;
}

export interface AssistanceSuggestion {
	id: string;
	session_id: string;
	type: 'suggested_response' | 'knowledge_article' | 'compliance_warning' | 'performance_tip' | 'escalation_guide';
	priority: 'low' | 'medium' | 'high' | 'urgent';
	title: string;
	content: string;
	confidence: number;
	timestamp: string;
	context: Record<string, unknown>;
}

export interface SentimentConfig {
	enable_real_time?: boolean;
	enable_emotions?: boolean;
	alert_on_negative?: boolean;
	negative_threshold?: number;
	track_trends?: boolean;
}

export interface SentimentAnalysis {
	id: string;
	session_id: string;
	sentiment: 'very_positive' | 'positive' | 'neutral' | 'negative' | 'very_negative';
	emotions: string[];
	confidence: number;
	polarity_score: number;
	intensity: number;
	timestamp: string;
	text_analyzed: string;
	speaker: string;
}

export interface SentimentTrend {
	session_id: string;
	current_sentiment: string;
	average_polarity: number;
	sentiment_changes: number;
	is_improving: boolean;
	alert_level: 'none' | 'low' | 'medium' | 'high';
}

// ============================================================================
// API Functions - Transcription
// ============================================================================

export async function startTranscription(sessionId: string, config: TranscriptionConfig): Promise<void> {
	const response = await fetch(`${API_BASE}/ai/transcription/start`, {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ session_id: sessionId, config })
	});

	if (!response.ok) {
		throw new Error(`Failed to start transcription: ${response.statusText}`);
	}
}

export async function stopTranscription(sessionId: string): Promise<void> {
	const response = await fetch(`${API_BASE}/ai/transcription/stop?session_id=${sessionId}`, {
		method: 'POST'
	});

	if (!response.ok) {
		throw new Error(`Failed to stop transcription: ${response.statusText}`);
	}
}

export async function addTranscription(
	sessionId: string,
	text: string,
	speaker: 'agent' | 'customer' | 'system',
	confidence = 1.0,
	isFinal = true
): Promise<TranscriptionSegment> {
	const response = await fetch(`${API_BASE}/ai/transcription/add`, {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({
			session_id: sessionId,
			text,
			speaker,
			confidence,
			is_final: isFinal
		})
	});

	if (!response.ok) {
		throw new Error(`Failed to add transcription: ${response.statusText}`);
	}

	return response.json();
}

export async function getTranscriptionHistory(sessionId: string, limit?: number): Promise<TranscriptionSegment[]> {
	const url = new URL(`${API_BASE}/ai/transcription/history/${sessionId}`);
	if (limit) url.searchParams.set('limit', limit.toString());

	const response = await fetch(url.toString());

	if (!response.ok) {
		throw new Error(`Failed to get transcription history: ${response.statusText}`);
	}

	return response.json();
}

export async function getFullTranscript(sessionId: string, includeInterim = false): Promise<{ session_id: string; transcript: string }> {
	const url = new URL(`${API_BASE}/ai/transcription/full/${sessionId}`);
	url.searchParams.set('include_interim', includeInterim.toString());

	const response = await fetch(url.toString());

	if (!response.ok) {
		throw new Error(`Failed to get full transcript: ${response.statusText}`);
	}

	return response.json();
}

export async function getTranscriptionStats(sessionId: string): Promise<Record<string, unknown>> {
	const response = await fetch(`${API_BASE}/ai/transcription/stats/${sessionId}`);

	if (!response.ok) {
		throw new Error(`Failed to get transcription stats: ${response.statusText}`);
	}

	return response.json();
}

// ============================================================================
// API Functions - Summarization
// ============================================================================

export async function generateSummary(sessionId: string, transcript: string, config: SummarizationConfig): Promise<CallSummary> {
	const response = await fetch(`${API_BASE}/ai/summarization/generate`, {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ session_id: sessionId, transcript, config })
	});

	if (!response.ok) {
		throw new Error(`Failed to generate summary: ${response.statusText}`);
	}

	return response.json();
}

export async function getSummary(sessionId: string): Promise<CallSummary> {
	const response = await fetch(`${API_BASE}/ai/summarization/${sessionId}`);

	if (!response.ok) {
		throw new Error(`Failed to get summary: ${response.statusText}`);
	}

	return response.json();
}

// ============================================================================
// API Functions - Agent Assistance
// ============================================================================

export async function startAssistance(sessionId: string, config: AssistanceConfig): Promise<void> {
	const response = await fetch(`${API_BASE}/ai/assistance/start`, {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ session_id: sessionId, config })
	});

	if (!response.ok) {
		throw new Error(`Failed to start assistance: ${response.statusText}`);
	}
}

export async function stopAssistance(sessionId: string): Promise<void> {
	const response = await fetch(`${API_BASE}/ai/assistance/stop?session_id=${sessionId}`, {
		method: 'POST'
	});

	if (!response.ok) {
		throw new Error(`Failed to stop assistance: ${response.statusText}`);
	}
}

export async function getAssistance(sessionId: string, transcript: string, context: Record<string, unknown> = {}): Promise<AssistanceSuggestion[]> {
	const response = await fetch(`${API_BASE}/ai/assistance/analyze`, {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ session_id: sessionId, transcript, context })
	});

	if (!response.ok) {
		throw new Error(`Failed to get assistance: ${response.statusText}`);
	}

	return response.json();
}

export async function getAssistanceHistory(sessionId: string, limit?: number): Promise<AssistanceSuggestion[]> {
	const url = new URL(`${API_BASE}/ai/assistance/history/${sessionId}`);
	if (limit) url.searchParams.set('limit', limit.toString());

	const response = await fetch(url.toString());

	if (!response.ok) {
		throw new Error(`Failed to get assistance history: ${response.statusText}`);
	}

	return response.json();
}

// ============================================================================
// API Functions - Sentiment Analysis
// ============================================================================

export async function startSentimentAnalysis(sessionId: string, config: SentimentConfig): Promise<void> {
	const response = await fetch(`${API_BASE}/ai/sentiment/start`, {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ session_id: sessionId, config })
	});

	if (!response.ok) {
		throw new Error(`Failed to start sentiment analysis: ${response.statusText}`);
	}
}

export async function stopSentimentAnalysis(sessionId: string): Promise<void> {
	const response = await fetch(`${API_BASE}/ai/sentiment/stop?session_id=${sessionId}`, {
		method: 'POST'
	});

	if (!response.ok) {
		throw new Error(`Failed to stop sentiment analysis: ${response.statusText}`);
	}
}

export async function analyzeSentiment(sessionId: string, text: string, speaker: string): Promise<SentimentAnalysis> {
	const response = await fetch(`${API_BASE}/ai/sentiment/analyze`, {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ session_id: sessionId, text, speaker })
	});

	if (!response.ok) {
		throw new Error(`Failed to analyze sentiment: ${response.statusText}`);
	}

	return response.json();
}

export async function getSentimentTrend(sessionId: string): Promise<SentimentTrend> {
	const response = await fetch(`${API_BASE}/ai/sentiment/trend/${sessionId}`);

	if (!response.ok) {
		throw new Error(`Failed to get sentiment trend: ${response.statusText}`);
	}

	return response.json();
}

export async function getSentimentHistory(sessionId: string, limit?: number): Promise<SentimentAnalysis[]> {
	const url = new URL(`${API_BASE}/ai/sentiment/history/${sessionId}`);
	if (limit) url.searchParams.set('limit', limit.toString());

	const response = await fetch(url.toString());

	if (!response.ok) {
		throw new Error(`Failed to get sentiment history: ${response.statusText}`);
	}

	return response.json();
}

export async function getSentimentAlerts(sessionId: string): Promise<Array<Record<string, unknown>>> {
	const response = await fetch(`${API_BASE}/ai/sentiment/alerts/${sessionId}`);

	if (!response.ok) {
		throw new Error(`Failed to get sentiment alerts: ${response.statusText}`);
	}

	return response.json();
}

File: previous-gen-apps/voice-kraliki/backend/app/api/ai_services.py
"""AI Services API endpoints.

Exposes transcription, summarization, agent assistance, and sentiment analysis
services via REST and WebSocket endpoints.
"""

import logging
from uuid import UUID

from fastapi import APIRouter, HTTPException, Request, WebSocket, WebSocketDisconnect
from pydantic import BaseModel

from app.middleware.rate_limit import AI_SERVICE_RATE_LIMIT, limiter
from app.services.agent_assistance_service import (
    AssistanceConfig,
    AssistanceSuggestion,
    get_agent_assistance_service,
)
from app.services.sentiment_service import (
    SentimentAnalysis,
    SentimentConfig,
    SentimentTrend,
    get_sentiment_service,
)
from app.services.summarization_service import (
    CallSummary,
    SummarizationConfig,
    get_summarization_service,
)
from app.services.transcription_service import (
    SpeakerRole,
    TranscriptionConfig,
    TranscriptionSegment,
    get_transcription_service,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/ai", tags=["AI Services"])


# Request/Response Models
class StartTranscriptionRequest(BaseModel):
    session_id: UUID
    config: TranscriptionConfig


class AddTranscriptionRequest(BaseModel):
    session_id: UUID
    text: str
    speaker: SpeakerRole
    confidence: float = 1.0
    is_final: bool = True


class GenerateSummaryRequest(BaseModel):
    session_id: UUID
    transcript: str
    config: SummarizationConfig


class AnalyzeSentimentRequest(BaseModel):
    session_id: UUID
    text: str
    speaker: str


class GetAssistanceRequest(BaseModel):
    session_id: UUID
    transcript: str
    context: dict = {}


# Transcription Endpoints
@router.post("/transcription/start")
@limiter.limit(AI_SERVICE_RATE_LIMIT)
async def start_transcription(http_request: Request, request: StartTranscriptionRequest):
    """Start transcription for a session. Rate limited (AI operation)."""
    try:
        service = get_transcription_service()
        await service.start_transcription(request.session_id, request.config)
        return {
            "status": "success",
            "message": f"Transcription started for session {request.session_id}"
        }
    except Exception as e:
        logger.error(f"Failed to start transcription: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/transcription/stop")
@limiter.limit(AI_SERVICE_RATE_LIMIT)
async def stop_transcription(request: Request, session_id: UUID):
    """Stop transcription for a session. Rate limited (AI operation)."""
    try:
        service = get_transcription_service()
        await service.stop_transcription(session_id)
        return {
            "status": "success",
            "message": f"Transcription stopped for session {session_id}"
        }
    except Exception as e:
        logger.error(f"Failed to stop transcription: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/transcription/add")
@limiter.limit(AI_SERVICE_RATE_LIMIT)
async def add_transcription(http_request: Request, request: AddTranscriptionRequest) -> TranscriptionSegment:
    """Add a transcription segment. Rate limited (AI operation)."""
    try:
        service = get_transcription_service()
        segment = await service.add_transcription(
            request.session_id,
            request.text,
            request.speaker,
            request.confidence,
            request.is_final
        )
        return segment
    except Exception as e:
        logger.error(f"Failed to add transcription: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/transcription/history/{session_id}")
async def get_transcription_history(
    session_id: UUID,
    limit: int | None = None
) -> list[TranscriptionSegment]:
    """Get transcription history for a session."""
    try:
        service = get_transcription_service()
        return service.get_transcription_history(session_id, limit)
    except Exception as e:
        logger.error(f"Failed to get transcription history: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/transcription/full/{session_id}")
async def get_full_transcript(
    session_id: UUID,
    include_interim: bool = False
) -> dict:
    """Get full transcript as formatted text."""
    try:
        service = get_transcription_service()
        transcript = service.get_full_transcript(session_id, include_interim)
        return {"session_id": str(session_id), "transcript": transcript}
    except Exception as e:
        logger.error(f"Failed to get full transcript: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/transcription/stats/{session_id}")
async def get_transcription_stats(session_id: UUID) -> dict:
    """Get transcription statistics."""
    try:
        service = get_transcription_service()
        return await service.get_session_stats(session_id)
    except Exception as e:
        logger.error(f"Failed to get transcription stats: {e}")
        raise HTTPException(status_code=500, detail=str(e))


# Summarization Endpoints
@router.post("/summarization/generate")
@limiter.limit(AI_SERVICE_RATE_LIMIT)
async def generate_summary(http_request: Request, request: GenerateSummaryRequest) -> CallSummary:
    """Generate AI-powered call summary. Rate limited (AI operation)."""
    try:
        service = get_summarization_service()
        summary = await service.generate_summary(
            request.session_id,
            request.transcript,
            request.config
        )
        return summary
    except Exception as e:
        logger.error(f"Failed to generate summary: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/summarization/{session_id}")
async def get_summary(session_id: UUID) -> CallSummary | None:
    """Get cached summary for a session."""
    try:
        service = get_summarization_service()
        summary = await service.get_summary(session_id)
        if not summary:
            raise HTTPException(status_code=404, detail="Summary not found")
        return summary
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to get summary: {e}")
        raise HTTPException(status_code=500, detail=str(e))


# Agent Assistance Endpoints
@router.post("/assistance/start")
@limiter.limit(AI_SERVICE_RATE_LIMIT)
async def start_assistance(request: Request, session_id: UUID, config: AssistanceConfig):
    """Start agent assistance for a session. Rate limited (AI operation)."""
    try:
        service = get_agent_assistance_service()
        await service.start_assistance(session_id, config)
        return {
            "status": "success",
            "message": f"Agent assistance started for session {session_id}"
        }
    except Exception as e:
        logger.error(f"Failed to start assistance: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/assistance/stop")
@limiter.limit(AI_SERVICE_RATE_LIMIT)
async def stop_assistance(request: Request, session_id: UUID):
    """Stop agent assistance for a session. Rate limited (AI operation)."""
    try:
        service = get_agent_assistance_service()
        await service.stop_assistance(session_id)
        return {
            "status": "success",
            "message": f"Agent assistance stopped for session {session_id}"
        }
    except Exception as e:
        logger.error(f"Failed to stop assistance: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/assistance/analyze")
@limiter.limit(AI_SERVICE_RATE_LIMIT)
async def get_assistance(http_request: Request, request: GetAssistanceRequest) -> list[AssistanceSuggestion]:
    """Get assistance suggestions for current conversation. Rate limited (AI operation)."""
    try:
        service = get_agent_assistance_service()
        suggestions = await service.analyze_conversation(
            request.session_id,
            request.transcript,
            request.context
        )
        return suggestions
    except Exception as e:
        logger.error(f"Failed to get assistance: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/assistance/history/{session_id}")
async def get_assistance_history(
    session_id: UUID,
    limit: int | None = None
) -> list[AssistanceSuggestion]:
    """Get assistance suggestion history."""
    try:
        service = get_agent_assistance_service()
        return service.get_suggestion_history(session_id, limit)
    except Exception as e:
        logger.error(f"Failed to get assistance history: {e}")
        raise HTTPException(status_code=500, detail=str(e))


# Sentiment Analysis Endpoints
@router.post("/sentiment/start")
@limiter.limit(AI_SERVICE_RATE_LIMIT)
async def start_sentiment_analysis(request: Request, session_id: UUID, config: SentimentConfig):
    """Start sentiment analysis for a session. Rate limited (AI operation)."""
    try:
        service = get_sentiment_service()
        await service.start_analysis(session_id, config)
        return {
            "status": "success",
            "message": f"Sentiment analysis started for session {session_id}"
        }
    except Exception as e:
        logger.error(f"Failed to start sentiment analysis: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/sentiment/stop")
@limiter.limit(AI_SERVICE_RATE_LIMIT)
async def stop_sentiment_analysis(request: Request, session_id: UUID):
    """Stop sentiment analysis for a session. Rate limited (AI operation)."""
    try:
        service = get_sentiment_service()
        await service.stop_analysis(session_id)
        return {
            "status": "success",
            "message": f"Sentiment analysis stopped for session {session_id}"
        }
    except Exception as e:
        logger.error(f"Failed to stop sentiment analysis: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/sentiment/analyze")
@limiter.limit(AI_SERVICE_RATE_LIMIT)
async def analyze_sentiment(http_request: Request, request: AnalyzeSentimentRequest) -> SentimentAnalysis | None:
    """Analyze sentiment of text. Rate limited (AI operation)."""
    try:
        service = get_sentiment_service()
        analysis = await service.analyze_text(
            request.session_id,
            request.text,
            request.speaker
        )
        return analysis
    except Exception as e:
        logger.error(f"Failed to analyze sentiment: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/sentiment/trend/{session_id}")
async def get_sentiment_trend(session_id: UUID) -> SentimentTrend | None:
    """Get sentiment trend for a session."""
    try:
        service = get_sentiment_service()
        trend = await service.get_sentiment_trend(session_id)
        if not trend:
            raise HTTPException(status_code=404, detail="No sentiment data found")
        return trend
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to get sentiment trend: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/sentiment/history/{session_id}")
async def get_sentiment_history(
    session_id: UUID,
    limit: int | None = None
) -> list[SentimentAnalysis]:
    """Get sentiment history for a session."""
    try:
        service = get_sentiment_service()
        return service.get_sentiment_history(session_id, limit)
    except Exception as e:
        logger.error(f"Failed to get sentiment history: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/sentiment/alerts/{session_id}")
async def get_sentiment_alerts(session_id: UUID) -> list[dict]:
    """Get sentiment alerts for a session."""
    try:
        service = get_sentiment_service()
        return service.get_alerts(session_id)
    except Exception as e:
        logger.error(f"Failed to get sentiment alerts: {e}")
        raise HTTPException(status_code=500, detail=str(e))


# WebSocket endpoint for real-time AI data
@router.websocket("/ws/{session_id}")
async def ai_websocket(websocket: WebSocket, session_id: UUID):
    """WebSocket endpoint for real-time AI service updates."""
    await websocket.accept()
    logger.info(f"AI WebSocket connected for session {session_id}")

    try:
        while True:
            # Receive messages from client
            data = await websocket.receive_json()
            message_type = data.get("type")

            if message_type == "transcription":
                # Add transcription
                service = get_transcription_service()
                segment = await service.add_transcription(
                    session_id,
                    data.get("text", ""),
                    SpeakerRole(data.get("speaker", "agent")),
                    data.get("confidence", 1.0),
                    data.get("is_final", True)
                )
                await websocket.send_json({
                    "type": "transcription",
                    "data": segment.dict()
                })

            elif message_type == "sentiment":
                # Analyze sentiment
                sentiment_service = get_sentiment_service()
                analysis = await sentiment_service.analyze_text(
                    session_id,
                    data.get("text", ""),
                    data.get("speaker", "customer")
                )
                if analysis:
                    await websocket.send_json({
                        "type": "sentiment",
                        "data": analysis.dict()
                    })

            elif message_type == "assistance":
                # Get assistance
                assistance_service = get_agent_assistance_service()
                suggestions = await assistance_service.analyze_conversation(
                    session_id,
                    data.get("transcript", ""),
                    data.get("context", {})
                )
                await websocket.send_json({
                    "type": "assistance",
                    "data": [s.dict() for s in suggestions]
                })

    except WebSocketDisconnect:
        logger.info(f"AI WebSocket disconnected for session {session_id}")
    except Exception as e:
        logger.error(f"AI WebSocket error for session {session_id}: {e}")
        await websocket.close()
